Frameworks/Starboard/PathMapper.cpp

```diff
@@ -0,0 +1,136 @@
+#include "StringHelpers.h"
+
+#include <algorithm>
+#include <list>
+#include <regex>
+
+#include "Platform/EbrPlatform.h"
+#include "PathMapper.h"
+
+// utility function to tokenize string using delmiters
+// d:\src/winobjc ==> d:, src, winobjc
+// /src/winobjc ==> "", src, winobjc
+// / ==> ""
+// src ==> "src"
+
+static const std::wstring c_currentDir(L".");
+static const std::wstring c_UpDir(L"..");
+
+static std::list<std::wstring> _TokenizeString(const std::wstring& str, const wchar_t* delims) {
+    std::list<std::wstring> components;
+
+    std::size_t start = 0; // start from 0
+    std::size_t delimPos = str.find_first_of(delims);
+
+    while (start != std::wstring::npos) {
+        components.emplace_back(str.substr(start, delimPos - start));
+        start = str.find_first_not_of(delims, delimPos);
+        delimPos = str.find_first_of(delims, start);
+    }
+    return components;
+}
+
+// converts any ? to +, we may add more later if needed for compat
+
+static void _EscapeIllegalPathCharacters(std::wstring& str) {
+    std::size_t pos = str.find_first_of(L'?');
+    while (pos != std::wstring::npos) {
+        str[pos] = L'+';
+        pos = str.find_first_of(L'?', pos + 1);
+    }
+}
+
+// normalize relative path
+//  1 remove any "." or "" components
+//  2 from reverse, remove any ".." and preceeding component
+// if components becomes empty, return "."
+
+static void _NormalizeRelativePathComponents(std::list<std::wstring>& components) {
+    for (auto it = components.begin(); it != components.end();) {
+        if (*it == c_currentDir || it->empty()) {
+            it = components.erase(it);
+            continue;
+        } else if (*it == c_UpDir) {
+            // remove previous component if it exists
+            if (it == components.begin()) {
+                // error, no previous directory exists
+                components.clear();
+                break;
+            }
+            // we have to do this in case the last component is ..
+            it = components.erase(--it);
+            it = components.erase(it);
+        } else {
+            ++it;
+        }
+    }
+
+    if (components.empty()) {
+        components.emplace_back(c_currentDir);
+    }
+}
+
+// the first component could have special meaning, we map it here
+static const std::wstring c_specialFolders[] = { L"Documents", L"cache", L"Library", L"AppSupport", L"tmp", L"shared" };
+
+std::wstring _MapPathRoot(const std::wstring& root) {
+    if (root == c_currentDir) {
+        return root;
+    } else if (root == L"~") {
+        return std::wstring(L".\\home");
+    }
+
+    std::wregex drive(L"[a-zA-Z]:");
+
+    if (std::regex_match(root, drive)) {
+        return root;
+    }
+
+    for (int i = 0; i < _countof(c_specialFolders); ++i) {
+        if (_wcsicmp(root.c_str(), c_specialFolders[i].c_str()) == 0) {
+            return Strings::NarrowToWide<std::wstring>(EbrGetWritableFolder()) + std::wstring(L"\\") + c_specialFolders[i];
```

Comment:
Is there room to exchange `EbrGetWritableFolder` with a new `_EbrGetWritableFolder`, its UTF-16 counterpart?